server/src/main/java/org/opensearch/gateway/TransportNodesListGatewayStartedShardsBatch.java

```diff
@@ -0,0 +1,404 @@
+
+        public NodesGatewayStartedShardsBatch(
+            ClusterName clusterName,
+            List<NodeGatewayStartedShardsBatch> nodes,
+            List<FailedNodeException> failures
+        ) {
+            super(clusterName, nodes, failures);
+        }
+
+        @Override
+        protected List<NodeGatewayStartedShardsBatch> readNodesFrom(StreamInput in) throws IOException {
+            return in.readList(NodeGatewayStartedShardsBatch::new);
+        }
+
+        @Override
+        protected void writeNodesTo(StreamOutput out, List<NodeGatewayStartedShardsBatch> nodes) throws IOException {
+            out.writeList(nodes);
+        }
+    }
+
+    /**
+     * NodeRequest class is for deserializing the  request received by this node from other node for this transport action.
+     * This is used in {@link TransportNodesAction}
+     *
+     * @opensearch.internal
+     */
+    public static class NodeRequest extends TransportRequest {
+        private final Map<ShardId, ShardAttributes> shardAttributes;
+
+        public NodeRequest(StreamInput in) throws IOException {
+            super(in);
+            shardAttributes = in.readMap(ShardId::new, ShardAttributes::new);
+        }
+
+        public NodeRequest(Request request) {
+            this.shardAttributes = Objects.requireNonNull(request.getShardAttributes());
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeMap(shardAttributes, (o, k) -> k.writeTo(o), (o, v) -> v.writeTo(o));
+        }
+    }
+
+    /**
+     * This class encapsulates the metadata about a started shard that needs to be persisted or sent between nodes.
+     * This is used in {@link NodeGatewayStartedShardsBatch} to construct the response for each node, instead of
+     * {@link TransportNodesListGatewayStartedShards.NodeGatewayStartedShards} because we don't need to save an extra
+     * {@link DiscoveryNode} object like in {@link TransportNodesListGatewayStartedShards.NodeGatewayStartedShards}
+     * which reduces memory footprint of its objects.
+     *
+     * @opensearch.internal
+     */
+    public static class NodeGatewayStartedShard {
+        private final String allocationId;
+        private final boolean primary;
+        private final Exception storeException;
+        private final ReplicationCheckpoint replicationCheckpoint;
+
+        public NodeGatewayStartedShard(StreamInput in) throws IOException {
+            allocationId = in.readOptionalString();
+            primary = in.readBoolean();
+            if (in.readBoolean()) {
+                storeException = in.readException();
+            } else {
+                storeException = null;
+            }
+            if (in.getVersion().onOrAfter(Version.V_2_3_0) && in.readBoolean()) {
+                replicationCheckpoint = new ReplicationCheckpoint(in);
+            } else {
+                replicationCheckpoint = null;
```

Comment:
we dont need this version check for batch request. It will always be after 2.3 right?

Reply:
Removed the replication check from the constructor of new Transport.